### PR TITLE
Windows: Use `FindFirstFileW` for getting the metadata of locked system files

### LINKED ISSUE #96980: [fs] Path::new(r#"C:\hiberfil.sys"#).exists() returns false, despite the file existing
I noticed this bug while researching the same bug in https://github.com/microsoft/STL.

```rust
// test.rs
use std::{env, fs};
use std::path::Path;

fn main() {
  for p in env::args().skip(1) {
    println!("{:?}.exists() = {}", p, Path::new(p).exists());
    println!("  => (metadata = {})", if fs::metadata(p).is_err() { "error" } else { "ok" });
  }
}
```

Given `test.exe C:\hiberfil.sys`, I expected:

```
"C:\\hiberfil.sys".exists() = true
  => (metadata = ok)
```

Instead, this happened:

```
"C:\\hiberfil.sys".exists() = false
  => (metadata = error)
```

### Solution

The related issue in the STL is microsoft/STL#2370; the PR to fix this in the STL is microsoft/STL#2715.

The problem is that `fs::metadata` opens the file; this is not allowed for the following four files (not sure if there are additional files which act similarly):

* `C:\DumpStack.log.tmp`
* `C:\hiberfil.sys`
* `C:\pagefile.sys`
* `C:\swapfile.sys`

We can get around this by using `FindFirstFileW` instead of `open()`; is that something that's desired, or is this considered a weird enough use case that we aren't interested in fixing it?

### PR DESCRIPTION
**Pull request by ChrisDenton**
Fixes #96980

Usually opening a file handle with access set to metadata only will always succeed, even if the file is locked. However some special system files, such as `C:\hiberfil.sys`, are locked by the system in a way that denies even that. So as a fallback we try reading the cached metadata from the directory.

Note that the test is a bit iffy. I don't know if `hiberfil.sys` actually exists in the CI.

r? rust-lang/libs

### CODE REVIEW

File: library/std/src/fs/tests.rs [hunk 1]
```diff
@@ -1534,3 +1534,14 @@ fn read_large_dir() {
         entry.unwrap();
     }
 }
+
+#[test]
+#[cfg(windows)]
+fn hiberfil_sys() {
+    // Get the system drive, which is usually `C:`.
```
Comment (bjorn3) [inline library/std/src/fs/tests.rs:1541]: I thought Windows has `C:` hard coded in various places and as such effectively requires the system drive to be `C:`.

File: library/std/src/fs/tests.rs [hunk 2]
```diff
@@ -1534,3 +1534,14 @@ fn read_large_dir() {
         entry.unwrap();
     }
 }
+
+#[test]
+#[cfg(windows)]
+fn hiberfil_sys() {
```
Comment (bjorn3) [inline library/std/src/fs/tests.rs:1540]: Maybe add a comment that this is testing the fallback method?

File: library/std/src/fs/tests.rs (hunk 1 above)
Comment (ChrisDenton) [reply library/std/src/fs/tests.rs:1541]: Pretty much. I didn't want to make any assumptions but I am probably being overly cautious in this case.

File: library/std/src/fs/tests.rs (hunk 2 above)
Comment (ChrisDenton) [reply library/std/src/fs/tests.rs:1540]: Will do.

**Comment by strega-nil-ms** [comment]
This is, as far as I can tell, equivalent to the code in the Microsoft STL: <https://github.com/microsoft/STL/blob/062002644137095c606f0a5a11d04f0278e8632e/stl/src/filesystem.cpp#L794>

**Review by thomcc — COMMENTED** [review]
Sorry it took so long to get to this. This looks good to me.

**Comment by ChrisDenton** [comment]
@bors r- failed in rollup

Oh, I was afraid of this. `hiberfil.sys` doesn't exist in CI. Hm, maybe we could keep the test but mark it as ignore? Admittedly it not running in CI very much limits its utility but at least it can be run locally.

**Comment by thomcc** [comment]
Oof, I thought about marking it as rollup=never for that reason (I guess doing a try run would have been more prudent...)

Marking the test as ignore makes sense to me.

**Comment by thomcc** [comment]
It seems a bit goofy to do a try run now that the test is ignored, so I'll just reapprove.

@bors r+

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/99412#issuecomment-1187567416)

@bors r-

**Comment by ChrisDenton** [comment]
Sorry, we already noticed that and addressed the issue.

@bors r=thomcc

**Pull request merged**